src/Compiler.Dynamic/SystemWebCompilation.cs

```diff
@@ -46,6 +46,11 @@ internal abstract class SystemWebCompilation<T> : IDisposable
         _logger = logger.CreateLogger<SystemWebCompilation<T>>();
         _pageCompilation = pageCompilation;
 
+        if (pageCompilation?.Value?.FileProvider is not null)
+        {
+            _env.ContentRootPath = pageCompilation.Value.FileProvider.Root;
```

Comment:
Instead of setting things here, can you register an options configuration here:

https://github.com/CoreWebForms/CoreWebForms/blob/2f514d65a2cb0802440d1fdadbff7ecdc4edd4e9/src/Compiler.Dynamic/WebFormsCompilerExtensions.cs#L46-L55

We don't want to update the IHostEnvironment value, but rather replace our usage of that value in the compiler with the new options you created

Reply:
Hi, I will look into it. Just out of curiosity - will changing IHostEnvironment value mess up static files, MVC, Razor Pages etc or is it more deeper? Thanks!

FYI, I had originally tried IFileProvider with the following which didn't work so went in the other direction:
```cs
public IFileProvider Files => _pageCompilation.Value.FileProvider ?? _env.ContentRootFileProvider;
```

Reply:
Looks like this will also have to change.

https://github.com/CoreWebForms/CoreWebForms/blob/f6dc6e712ec35c05578ece295f4daca298af9278/src/WebForms/VirtualPath.cs#L13

Reply:
Yes, changing the host env itself will affect the rest of the stack.

Good catch on the virtual file - search for other usages of ihostenvironment or iwebhostenviroent (there are similar apis I don't think I've been consistent in using)

Reply:
It seems like we need WebFormsOptions that is passed to AddWebForms allowing you to set the RootDirectory like Razor Pages does. The default can be "/". This way, the aspx files organization be consistent between compiled and dynamic compiled mode.

Reply:
Hmmm... Not sure I understand how it would work for the static compilation. For that, we'd probably want an msbuild property - how do you see it working?

Reply:
Maybe I am making too many assumptions. Razor Pages are statically compiled and it only serves things inside Pages folder. You can also change to a different folder by setting RazorPageOptions.RootDirectory so I assumed we can do something similar.

https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.RazorPages/src/RazorPagesOptions.cs#L28

Reply:
In any case, want to validate my thought process. To customize VirtualPath I imagine we need to set a new IFileProvider in the WebForms project, then VirtualPath can use the new IFileProvider. Since VirtualPath is used both by dynamic and static compilation, the IFileProvider should be custom if one is set else default to env.ContentRootFileProvider.

Reply:
I get what you're describing, but the directory for the static compilation would have to be configured for the build system (ie msbuild) to be aware of. Also, I don't believe you can configure the static compilation of razor files to be in a different directory (see https://learn.microsoft.com/enus/aspnet/core/mvc/views/view-compilation?view=aspnetcore-8.0&tabs=visual-studio). 

I would suggest keeping things scoped so you don't have feature creep; i.e.

- [ ] start with enabling it for dynamic compilation with a custom IFileProvider options entry
- [ ] get it working for static compilation
- [ ] unify the api for that

I'm on board for the 1st one, but (2)/(3) I'm not sure i understand the how of it completely. Feel free to explore it more, though

Reply:
ah I get more of the issue - it's a layering thing of where VirtualPath is. For now, adding a WebFormsOptions type that has a content root that defaults the env.ContentRootFileProvider would make sense.

I'd like to not expose APIs off of the system.web types for now that did not exist before. they were not designed well for di or injecting new behavior so defining new apis on top that internally hook things up would be better